include/deal.II/matrix_free/cuda_tensor_product_kernels.h

```diff
@@ -298,42 +367,37 @@ namespace CUDAWrappers
 
 
     template <int dim, int fe_degree, int n_q_points_1d, typename Number>
-    inline __device__ void
+    template <typename ViewType>
+    DEAL_II_HOST_DEVICE inline void
     EvaluatorTensorProduct<evaluate_general,
                            dim,
                            fe_degree,
                            n_q_points_1d,
-                           Number>::integrate_value(Number *u)
+                           Number>::integrate_value(ViewType u)
     {
       switch (dim)
         {
           case 1:
             {
-              values<0, false, false, true>(
-                get_global_shape_values<Number>(mf_object_id), u, u);
+              values<0, false, false, true>(u, u);
 
               break;
             }
           case 2:
             {
-              values<0, false, false, true>(
-                get_global_shape_values<Number>(mf_object_id), u, u);
-              __syncthreads();
-              values<1, false, false, true>(
-                get_global_shape_values<Number>(mf_object_id), u, u);
+              values<0, false, false, true>(u, u);
+              KOKKOS_IF_ON_DEVICE(__syncthreads();)
+              values<1, false, false, true>(u, u);
 
               break;
             }
           case 3:
             {
-              values<0, false, false, true>(
-                get_global_shape_values<Number>(mf_object_id), u, u);
-              __syncthreads();
-              values<1, false, false, true>(
-                get_global_shape_values<Number>(mf_object_id), u, u);
-              __syncthreads();
-              values<2, false, false, true>(
-                get_global_shape_values<Number>(mf_object_id), u, u);
+              values<0, false, false, true>(u, u);
+              KOKKOS_IF_ON_DEVICE(__syncthreads();)
```

Comment:
These macros are really odd...

Reply:
They are temporary. We need them for now because we have a mix of CUDA and Kokkos. Once I remove the CUDA code they will disappear.